In [505]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/titanic/train.csv
/kaggle/input/titanic/test.csv


In [506]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

In [507]:
from sklearn.linear_model import LogisticRegression

In [508]:
df = pd.read_csv('../input/titanic/train.csv')

In [509]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [510]:
df.head(20)

PassengerId  Survived  Pclass  \
0             1         0       3   
1             2         1       1   
2             3         1       3   
3             4         1       1   
4             5         0       3   
5             6         0       3   
6             7         0       1   
7             8         0       3   
8             9         1       3   
9            10         1       2   
10           11         1       3   
11           12         1       1   
12           13         0       3   
13           14         0       3   
14           15         0       3   
15           16         1       2   
16           17         0       3   
17           18         1       2   
18           19         0       3   
19           20         1       3   

                                                 Name     Sex   Age  SibSp  \
0                             Braund, Mr. Owen Harris    male  22.0      1   
1   Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                              Heikkinen, Miss. Laina  female  26.0      0   
3        Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                            Allen, Mr. William Henry    male  35.0      0   
5                                    Moran, Mr. James    male   NaN      0   
6                             McCarthy, Mr. Timothy J    male  54.0      0   
7                      Palsson, Master. Gosta Leonard    male   2.0      3   
8   Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                 Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
10                    Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                           Bonnell, Miss. Elizabeth  female  58.0      0   
12                     Saundercock, Mr. William Henry    male  20.0      0   
13                        Andersson, Mr. Anders Johan    male  39.0      1   
14               Vestrom, Miss. Hulda Amanda Adolfina  female  14.0      0   
15                   Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
16                               Rice, Master. Eugene    male   2.0      4   
17                       Williams, Mr. Charles Eugene    male   NaN      0   
18  Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0      1   
19                            Masselmani, Mrs. Fatima  female   NaN      0   

    Parch            Ticket     Fare Cabin Embarked  
0       0         A/5 21171   7.2500   NaN        S  
1       0          PC 17599  71.2833   C85        C  
2       0  STON/O2. 3101282   7.9250   NaN        S  
3       0            113803  53.1000  C123        S  
4       0            373450   8.0500   NaN        S  
5       0            330877   8.4583   NaN        Q  
6       0             17463  51.8625   E46        S  
7       1            349909  21.0750   NaN        S  
8       2            347742  11.1333   NaN        S  
9       0            237736  30.0708   NaN        C  
10      1           PP 9549  16.7000    G6        S  
11      0            113783  26.5500  C103        S  
12      0         A/5. 2151   8.0500   NaN        S  
13      5            347082  31.2750   NaN        S  
14      0            350406   7.8542   NaN        S  
15      0            248706  16.0000   NaN        S  
16      1            382652  29.1250   NaN        Q  
17      0            244373  13.0000   NaN        S  
18      0            345763  18.0000   NaN        S  
19      0              2649   7.2250   NaN        C

# Initial thoughts on each field:
* 1) PassengerId  891 non-null    int64  - Meaningless key: delete
* 2) Survived     891 non-null    int64  - The Boolean target to predict
* 3) Pclass       891 non-null    int64  - 1, 2, 3. Proxy for social class, so use as an integer
* 4) Name         891 non-null    object - Scope for feature engineering a person's title, also Swedish (lots of passengers) names often end in _sen
* 5) Sex          891 non-null    object - Simple male/female. No non-binary in 1912!
* 6) Age          714 non-null    float64- 80% present. Worth imputing the missing 20%
* 7) SibSp        891 non-null    int64  - Explore whether families all die/survive together
* 8) Parch        891 non-null    int64  - Explore whether families all die/survive together
* 9) Ticket       891 non-null    object - I can't imagine the ticket code has any importance: delete
* 10) Fare         891 non-null    float64- Originally LSD, but converted to decimal. Multiples of guineas (£1.05) might show upper class
* 11) Cabin        204 non-null    object - 23% present. First letter probably shows the deck, ie proximity to lifeboat
* 12) Embarked     889 non-null    object - Possible discrimination against Irish passengers

In [511]:
print(df['Survived'].sum()/891*100, '% survival rate')

38.38383838383838 % survival rate


# 3) Passenger Class
1, 2 or 3. 
Proxy for social class, so use it as an integer, not categorical Booleans

In [512]:
df['Pclass'].unique()

array([3, 1, 2])

In [513]:
passengers_by_class = df.groupby('Pclass').count()['PassengerId']

In [514]:
passengers_by_class

Pclass
1    216
2    184
3    491
Name: PassengerId, dtype: int64

In [515]:
survival_by_class = df.groupby('Pclass').sum()['Survived']

In [516]:
survival_by_class

Pclass
1    136
2     87
3    119
Name: Survived, dtype: int64

In [517]:
survival_by_class['total']=passengers_by_class

In [518]:
survival_by_class

Pclass
1                                                      136
2                                                       87
3                                                      119
total    Pclass
1    216
2    184
3    491
Name: Passen...
Name: Survived, dtype: object

In [519]:
survival_rate_by_class = np.divide(survival_by_class, passengers_by_class)

In [520]:
survival_rate_by_class

Pclass
1                                                 0.629630
2                                                 0.472826
3                                                 0.242363
total    Pclass
1   NaN
2   NaN
3   NaN
Name: Passenger...
dtype: object

## No surprise: negative correlation between class (number) and survival

# 4) Name

Scope for feature engineering a person's title. <br>
I read somewhere that there were lots of Swedish passengers, whose names often end in _sen <br>
Format: Surname, Title Forenames eg Tart, Mr Nigel

In [521]:
#df[['surame', 'title_forenames']] = 

#df[['surname', 'title', 'forenames']] = 
df['Name'].str.split(' ', expand=True)

0      1          2        3          4       5        6     7   \
0       Braund,    Mr.       Owen   Harris       None    None     None  None   
1      Cumings,   Mrs.       John  Bradley  (Florence  Briggs  Thayer)  None   
2    Heikkinen,  Miss.      Laina     None       None    None     None  None   
3     Futrelle,   Mrs.    Jacques    Heath      (Lily     May    Peel)  None   
4        Allen,    Mr.    William    Henry       None    None     None  None   
..          ...    ...        ...      ...        ...     ...      ...   ...   
886   Montvila,   Rev.     Juozas     None       None    None     None  None   
887     Graham,  Miss.   Margaret    Edith       None    None     None  None   
888   Johnston,  Miss.  Catherine    Helen   "Carrie"    None     None  None   
889       Behr,    Mr.       Karl   Howell       None    None     None  None   
890     Dooley,    Mr.    Patrick     None       None    None     None  None   

       8     9     10    11    12    13  
0    None  None  None  None  None  None  
1    None  None  None  None  None  None  
2    None  None  None  None  None  None  
3    None  None  None  None  None  None  
4    None  None  None  None  None  None  
..    ...   ...   ...   ...   ...   ...  
886  None  None  None  None  None  None  
887  None  None  None  None  None  None  
888  None  None  None  None  None  None  
889  None  None  None  None  None  None  
890  None  None  None  None  None  None  

[891 rows x 14 columns]

In [522]:
df['Title'] = df['Name'].str.split(',').str[1].str.split('.').str[0].str.strip()

In [523]:
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked Title  
0        0         A/5 21171   7.2500   NaN        S    Mr  
1        0          PC 17599  71.2833   C85        C   Mrs  
2        0  STON/O2. 3101282   7.9250   NaN        S  Miss  
3        0            113803  53.1000  C123        S   Mrs  
4        0            373450   8.0500   NaN        S    Mr  
..     ...               ...      ...   ...      ...   ...  
886      0            211536  13.0000   NaN        S   Rev  
887      0            112053  30.0000   B42        S  Miss  
888      2        W./C. 6607  23.4500   NaN        S  Miss  
889      0            111369  30.0000  C148        C    Mr  
890      0            370376   7.7500   NaN        Q    Mr  

[891 rows x 13 columns]

In [524]:
df['Title'].unique()

array(['Mr', 'Mrs', 'Miss', 'Master', 'Don', 'Rev', 'Dr', 'Mme', 'Ms',
       'Major', 'Lady', 'Sir', 'Mlle', 'Col', 'Capt', 'the Countess',
       'Jonkheer'], dtype=object)

In [525]:
passengers_by_title = df.groupby('Title').count()['PassengerId']

In [526]:
passengers_by_title

Title
Capt              1
Col               2
Don               1
Dr                7
Jonkheer          1
Lady              1
Major             2
Master           40
Miss            182
Mlle              2
Mme               1
Mr              517
Mrs             125
Ms                1
Rev               6
Sir               1
the Countess      1
Name: PassengerId, dtype: int64

## Only the common English titles Mr, Mrs, Miss, Master have significant numbers (but comprising the vast majority of passengers). 

In [527]:
survival_by_title = df.groupby('Title').sum()['Survived']

In [528]:
survival_rate_by_title = np.divide(survival_by_title, passengers_by_title)

In [529]:
survival_rate_by_title

Title
Capt            0.000000
Col             0.500000
Don             0.000000
Dr              0.428571
Jonkheer        0.000000
Lady            1.000000
Major           0.500000
Master          0.575000
Miss            0.697802
Mlle            1.000000
Mme             1.000000
Mr              0.156673
Mrs             0.792000
Ms              1.000000
Rev             0.000000
Sir             1.000000
the Countess    1.000000
dtype: float64

In [530]:
df['IsMrs'] = df['Title'] == 'Mrs'

In [531]:
df['IsMr'] = df['Title'] == 'Mr'

In [532]:
df['IsMiss'] = df['Title'] == 'Miss'

In [533]:
df['IsMaster'] = df['Title'] == 'Master'

In [534]:
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked Title  IsMrs   IsMr  \
0        0         A/5 21171   7.2500   NaN        S    Mr  False   True   
1        0          PC 17599  71.2833   C85        C   Mrs   True  False   
2        0  STON/O2. 3101282   7.9250   NaN        S  Miss  False  False   
3        0            113803  53.1000  C123        S   Mrs   True  False   
4        0            373450   8.0500   NaN        S    Mr  False   True   
..     ...               ...      ...   ...      ...   ...    ...    ...   
886      0            211536  13.0000   NaN        S   Rev  False  False   
887      0            112053  30.0000   B42        S  Miss  False  False   
888      2        W./C. 6607  23.4500   NaN        S  Miss  False  False   
889      0            111369  30.0000  C148        C    Mr  False   True   
890      0            370376   7.7500   NaN        Q    Mr  False   True   

     IsMiss  IsMaster  
0     False     False  
1     False     False  
2      True     False  
3     False     False  
4     False     False  
..      ...       ...  
886   False     False  
887    True     False  
888    True     False  
889   False     False  
890   False     False  

[891 rows x 17 columns]

In [535]:
df = df.drop('Title', axis=1)
df = df.drop('Name', axis=1)

In [536]:
df

PassengerId  Survived  Pclass     Sex   Age  SibSp  Parch  \
0              1         0       3    male  22.0      1      0   
1              2         1       1  female  38.0      1      0   
2              3         1       3  female  26.0      0      0   
3              4         1       1  female  35.0      1      0   
4              5         0       3    male  35.0      0      0   
..           ...       ...     ...     ...   ...    ...    ...   
886          887         0       2    male  27.0      0      0   
887          888         1       1  female  19.0      0      0   
888          889         0       3  female   NaN      1      2   
889          890         1       1    male  26.0      0      0   
890          891         0       3    male  32.0      0      0   

               Ticket     Fare Cabin Embarked  IsMrs   IsMr  IsMiss  IsMaster  
0           A/5 21171   7.2500   NaN        S  False   True   False     False  
1            PC 17599  71.2833   C85        C   True  False   False     False  
2    STON/O2. 3101282   7.9250   NaN        S  False  False    True     False  
3              113803  53.1000  C123        S   True  False   False     False  
4              373450   8.0500   NaN        S  False   True   False     False  
..                ...      ...   ...      ...    ...    ...     ...       ...  
886            211536  13.0000   NaN        S  False  False   False     False  
887            112053  30.0000   B42        S  False  False    True     False  
888        W./C. 6607  23.4500   NaN        S  False  False    True     False  
889            111369  30.0000  C148        C  False   True   False     False  
890            370376   7.7500   NaN        Q  False   True   False     False  

[891 rows x 15 columns]

# 5) Sex

In [537]:
df['IsMale'] = df['Sex']=='Male'

In [538]:
df.drop(['Sex'], axis=1, inplace=True)

Simple male/female. No non-binary in 1912!

# 6) Age

80% present. Worth imputing the missing 20% <br>
Use mean age for Mr, Mrs, Miss, Master, then overall mean age for others

In [539]:
grouped_by_title = df.groupby(['IsMrs', 'IsMr', 'IsMiss', 'IsMaster'])

In [540]:
mean_age_by_title = grouped_by_title['Age'].mean()

In [541]:
# mean_age_by_title is joined to df on the Boolean fields
# mean_age_by_title also has a field called 'Age', so use suffixes to append '_mean' and make it distinct
df2 = pd.merge(df, mean_age_by_title, on=['IsMrs', 'IsMr', 'IsMiss', 'IsMaster'], suffixes = ('', '_mean')) 

In [542]:
df2

PassengerId  Survived  Pclass   Age  SibSp  Parch     Ticket     Fare  \
0              1         0       3  22.0      1      0  A/5 21171   7.2500   
1              5         0       3  35.0      0      0     373450   8.0500   
2              6         0       3   NaN      0      0     330877   8.4583   
3              7         0       1  54.0      0      0      17463  51.8625   
4             13         0       3  20.0      0      0  A/5. 2151   8.0500   
..           ...       ...     ...   ...    ...    ...        ...      ...   
886          767         0       1   NaN      0      0     112379  39.6000   
887          797         1       1  49.0      0      0      17465  25.9292   
888          823         0       1  38.0      0      0      19972   0.0000   
889          849         0       2  28.0      0      1     248727  33.0000   
890          887         0       2  27.0      0      0     211536  13.0000   

    Cabin Embarked  IsMrs   IsMr IsMiss IsMaster  IsMale   Age_mean  
0     NaN        S  False   True  False    False   False  32.368090  
1     NaN        S  False   True  False    False   False  32.368090  
2     NaN        Q  False   True  False    False   False  32.368090  
3     E46        S  False   True  False    False   False  32.368090  
4     NaN        S  False   True  False    False   False  32.368090  
..    ...      ...    ...    ...    ...      ...     ...        ...  
886   NaN        C  False  False  False    False   False  42.384615  
887   D17        S  False  False  False    False   False  42.384615  
888   NaN        S  False  False  False    False   False  42.384615  
889   NaN        S  False  False  False    False   False  42.384615  
890   NaN        S  False  False  False    False   False  42.384615  

[891 rows x 16 columns]

In [543]:
# Now replace the missing ages with the values of Age_mean
df2['Age'].fillna(df2['Age_mean'], inplace=True)

In [544]:
df = df2.drop(['Age_mean'], axis=1)

In [545]:
df

PassengerId  Survived  Pclass        Age  SibSp  Parch     Ticket  \
0              1         0       3  22.000000      1      0  A/5 21171   
1              5         0       3  35.000000      0      0     373450   
2              6         0       3  32.368090      0      0     330877   
3              7         0       1  54.000000      0      0      17463   
4             13         0       3  20.000000      0      0  A/5. 2151   
..           ...       ...     ...        ...    ...    ...        ...   
886          767         0       1  42.384615      0      0     112379   
887          797         1       1  49.000000      0      0      17465   
888          823         0       1  38.000000      0      0      19972   
889          849         0       2  28.000000      0      1     248727   
890          887         0       2  27.000000      0      0     211536   

        Fare Cabin Embarked  IsMrs   IsMr IsMiss IsMaster  IsMale  
0     7.2500   NaN        S  False   True  False    False   False  
1     8.0500   NaN        S  False   True  False    False   False  
2     8.4583   NaN        Q  False   True  False    False   False  
3    51.8625   E46        S  False   True  False    False   False  
4     8.0500   NaN        S  False   True  False    False   False  
..       ...   ...      ...    ...    ...    ...      ...     ...  
886  39.6000   NaN        C  False  False  False    False   False  
887  25.9292   D17        S  False  False  False    False   False  
888   0.0000   NaN        S  False  False  False    False   False  
889  33.0000   NaN        S  False  False  False    False   False  
890  13.0000   NaN        S  False  False  False    False   False  

[891 rows x 15 columns]

# 7) Sibling/Spouse

Explore whether families all die/survive together

Actually, the data doesn't show whether families die together; just whether having a sibling/spouse correlates with survival/death

The data is already integers, so ready to go

# 8) Parent/Child

Explore whether families all die/survive together

As above. Ready to go

# 9) Ticket Code

I can't imagine the ticket code has any importance: delete

In [546]:
df.drop(['Ticket'], axis=1, inplace=True)

In [547]:
df

PassengerId  Survived  Pclass        Age  SibSp  Parch     Fare Cabin  \
0              1         0       3  22.000000      1      0   7.2500   NaN   
1              5         0       3  35.000000      0      0   8.0500   NaN   
2              6         0       3  32.368090      0      0   8.4583   NaN   
3              7         0       1  54.000000      0      0  51.8625   E46   
4             13         0       3  20.000000      0      0   8.0500   NaN   
..           ...       ...     ...        ...    ...    ...      ...   ...   
886          767         0       1  42.384615      0      0  39.6000   NaN   
887          797         1       1  49.000000      0      0  25.9292   D17   
888          823         0       1  38.000000      0      0   0.0000   NaN   
889          849         0       2  28.000000      0      1  33.0000   NaN   
890          887         0       2  27.000000      0      0  13.0000   NaN   

    Embarked  IsMrs   IsMr IsMiss IsMaster  IsMale  
0          S  False   True  False    False   False  
1          S  False   True  False    False   False  
2          Q  False   True  False    False   False  
3          S  False   True  False    False   False  
4          S  False   True  False    False   False  
..       ...    ...    ...    ...      ...     ...  
886        C  False  False  False    False   False  
887        S  False  False  False    False   False  
888        S  False  False  False    False   False  
889        S  False  False  False    False   False  
890        S  False  False  False    False   False  

[891 rows x 14 columns]

# 10) Fare

Originally LSD, but converted to decimal. Multiples of guineas (£1.05) might show upper class

In [548]:
df['IsGuineas'] = [(fare % 1.05) == 0 for fare in df['Fare']]

In [549]:
df

PassengerId  Survived  Pclass        Age  SibSp  Parch     Fare Cabin  \
0              1         0       3  22.000000      1      0   7.2500   NaN   
1              5         0       3  35.000000      0      0   8.0500   NaN   
2              6         0       3  32.368090      0      0   8.4583   NaN   
3              7         0       1  54.000000      0      0  51.8625   E46   
4             13         0       3  20.000000      0      0   8.0500   NaN   
..           ...       ...     ...        ...    ...    ...      ...   ...   
886          767         0       1  42.384615      0      0  39.6000   NaN   
887          797         1       1  49.000000      0      0  25.9292   D17   
888          823         0       1  38.000000      0      0   0.0000   NaN   
889          849         0       2  28.000000      0      1  33.0000   NaN   
890          887         0       2  27.000000      0      0  13.0000   NaN   

    Embarked  IsMrs   IsMr IsMiss IsMaster  IsMale  IsGuineas  
0          S  False   True  False    False   False      False  
1          S  False   True  False    False   False      False  
2          Q  False   True  False    False   False      False  
3          S  False   True  False    False   False      False  
4          S  False   True  False    False   False      False  
..       ...    ...    ...    ...      ...     ...        ...  
886        C  False  False  False    False   False      False  
887        S  False  False  False    False   False      False  
888        S  False  False  False    False   False       True  
889        S  False  False  False    False   False      False  
890        S  False  False  False    False   False      False  

[891 rows x 15 columns]

In [550]:
# check out those fares which are multipleas of a Guinea - the only one in the first/last five is actually zero
df[df['IsGuineas'] == True]

PassengerId  Survived  Pclass       Age  SibSp  Parch  Fare Cabin  \
101          180         0       3  36.00000      0      0   0.0   NaN   
148          264         0       1  40.00000      0      0   0.0   B94   
153          272         1       3  25.00000      0      0   0.0   NaN   
155          278         0       2  32.36809      0      0   0.0   NaN   
171          303         0       3  19.00000      0      0   0.0   NaN   
226          414         0       2  32.36809      0      0   0.0   NaN   
259          467         0       2  32.36809      0      0   0.0   NaN   
268          482         0       2  32.36809      0      0   0.0   NaN   
338          598         0       3  49.00000      0      0   0.0   NaN   
362          634         0       1  32.36809      0      0   0.0   NaN   
388          675         0       2  32.36809      0      0   0.0   NaN   
426          733         0       2  32.36809      0      0   0.0   NaN   
471          807         0       1  39.00000      0      0   0.0   A36   
477          816         0       1  32.36809      0      0   0.0  B102   
888          823         0       1  38.00000      0      0   0.0   NaN   

    Embarked  IsMrs   IsMr IsMiss IsMaster  IsMale  IsGuineas  
101        S  False   True  False    False   False       True  
148        S  False   True  False    False   False       True  
153        S  False   True  False    False   False       True  
155        S  False   True  False    False   False       True  
171        S  False   True  False    False   False       True  
226        S  False   True  False    False   False       True  
259        S  False   True  False    False   False       True  
268        S  False   True  False    False   False       True  
338        S  False   True  False    False   False       True  
362        S  False   True  False    False   False       True  
388        S  False   True  False    False   False       True  
426        S  False   True  False    False   False       True  
471        S  False   True  False    False   False       True  
477        S  False   True  False    False   False       True  
888        S  False  False  False    False   False       True

Nice idea, but all the multiples of £1.05 are actually zero. <br>
Interestingly, all those passengers are male (14 Mr and 1 other title)

In [551]:
df.drop(['IsGuineas'], axis=1, inplace=True)

In [552]:
df

PassengerId  Survived  Pclass        Age  SibSp  Parch     Fare Cabin  \
0              1         0       3  22.000000      1      0   7.2500   NaN   
1              5         0       3  35.000000      0      0   8.0500   NaN   
2              6         0       3  32.368090      0      0   8.4583   NaN   
3              7         0       1  54.000000      0      0  51.8625   E46   
4             13         0       3  20.000000      0      0   8.0500   NaN   
..           ...       ...     ...        ...    ...    ...      ...   ...   
886          767         0       1  42.384615      0      0  39.6000   NaN   
887          797         1       1  49.000000      0      0  25.9292   D17   
888          823         0       1  38.000000      0      0   0.0000   NaN   
889          849         0       2  28.000000      0      1  33.0000   NaN   
890          887         0       2  27.000000      0      0  13.0000   NaN   

    Embarked  IsMrs   IsMr IsMiss IsMaster  IsMale  
0          S  False   True  False    False   False  
1          S  False   True  False    False   False  
2          Q  False   True  False    False   False  
3          S  False   True  False    False   False  
4          S  False   True  False    False   False  
..       ...    ...    ...    ...      ...     ...  
886        C  False  False  False    False   False  
887        S  False  False  False    False   False  
888        S  False  False  False    False   False  
889        S  False  False  False    False   False  
890        S  False  False  False    False   False  

[891 rows x 14 columns]

# 11) Cabin Number

Only 23% present, so probably drop the field, but the first letter probably shows the deck, ie proximity to lifeboat

In [553]:
df.drop(['Cabin'], axis=1, inplace=True)

In [554]:
df

PassengerId  Survived  Pclass        Age  SibSp  Parch     Fare Embarked  \
0              1         0       3  22.000000      1      0   7.2500        S   
1              5         0       3  35.000000      0      0   8.0500        S   
2              6         0       3  32.368090      0      0   8.4583        Q   
3              7         0       1  54.000000      0      0  51.8625        S   
4             13         0       3  20.000000      0      0   8.0500        S   
..           ...       ...     ...        ...    ...    ...      ...      ...   
886          767         0       1  42.384615      0      0  39.6000        C   
887          797         1       1  49.000000      0      0  25.9292        S   
888          823         0       1  38.000000      0      0   0.0000        S   
889          849         0       2  28.000000      0      1  33.0000        S   
890          887         0       2  27.000000      0      0  13.0000        S   

     IsMrs   IsMr IsMiss IsMaster  IsMale  
0    False   True  False    False   False  
1    False   True  False    False   False  
2    False   True  False    False   False  
3    False   True  False    False   False  
4    False   True  False    False   False  
..     ...    ...    ...      ...     ...  
886  False  False  False    False   False  
887  False  False  False    False   False  
888  False  False  False    False   False  
889  False  False  False    False   False  
890  False  False  False    False   False  

[891 rows x 13 columns]

# 12) Place of Embarkation

Queenstown (now Cobh) is in Irealand. Possible discrimination against Irish passengers

In [555]:
df['Southampton'] = df['Embarked'] == 'S'
df['Cherbourg'] = df['Embarked'] == 'C'
# All others are Queenstown (where S and C are False)

In [556]:
df.drop(['Embarked'], axis=1, inplace=True)

In [557]:
df

PassengerId  Survived  Pclass        Age  SibSp  Parch     Fare  IsMrs  \
0              1         0       3  22.000000      1      0   7.2500  False   
1              5         0       3  35.000000      0      0   8.0500  False   
2              6         0       3  32.368090      0      0   8.4583  False   
3              7         0       1  54.000000      0      0  51.8625  False   
4             13         0       3  20.000000      0      0   8.0500  False   
..           ...       ...     ...        ...    ...    ...      ...    ...   
886          767         0       1  42.384615      0      0  39.6000  False   
887          797         1       1  49.000000      0      0  25.9292  False   
888          823         0       1  38.000000      0      0   0.0000  False   
889          849         0       2  28.000000      0      1  33.0000  False   
890          887         0       2  27.000000      0      0  13.0000  False   

      IsMr IsMiss IsMaster  IsMale  Southampton  Cherbourg  
0     True  False    False   False         True      False  
1     True  False    False   False         True      False  
2     True  False    False   False        False      False  
3     True  False    False   False         True      False  
4     True  False    False   False         True      False  
..     ...    ...      ...     ...          ...        ...  
886  False  False    False   False        False       True  
887  False  False    False   False         True      False  
888  False  False    False   False         True      False  
889  False  False    False   False         True      False  
890  False  False    False   False         True      False  

[891 rows x 14 columns]

# 1) PassengerId

In [558]:
df.drop(['PassengerId'], axis=1, inplace=True)

In [559]:
df

Survived  Pclass        Age  SibSp  Parch     Fare  IsMrs   IsMr IsMiss  \
0           0       3  22.000000      1      0   7.2500  False   True  False   
1           0       3  35.000000      0      0   8.0500  False   True  False   
2           0       3  32.368090      0      0   8.4583  False   True  False   
3           0       1  54.000000      0      0  51.8625  False   True  False   
4           0       3  20.000000      0      0   8.0500  False   True  False   
..        ...     ...        ...    ...    ...      ...    ...    ...    ...   
886         0       1  42.384615      0      0  39.6000  False  False  False   
887         1       1  49.000000      0      0  25.9292  False  False  False   
888         0       1  38.000000      0      0   0.0000  False  False  False   
889         0       2  28.000000      0      1  33.0000  False  False  False   
890         0       2  27.000000      0      0  13.0000  False  False  False   

    IsMaster  IsMale  Southampton  Cherbourg  
0      False   False         True      False  
1      False   False         True      False  
2      False   False        False      False  
3      False   False         True      False  
4      False   False         True      False  
..       ...     ...          ...        ...  
886    False   False        False       True  
887    False   False         True      False  
888    False   False         True      False  
889    False   False         True      False  
890    False   False         True      False  

[891 rows x 13 columns]

In [560]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 891 entries, 0 to 890
Data columns (total 13 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Survived     891 non-null    int64  
 1   Pclass       891 non-null    int64  
 2   Age          891 non-null    float64
 3   SibSp        891 non-null    int64  
 4   Parch        891 non-null    int64  
 5   Fare         891 non-null    float64
 6   IsMrs        891 non-null    object 
 7   IsMr         891 non-null    object 
 8   IsMiss       891 non-null    object 
 9   IsMaster     891 non-null    object 
 10  IsMale       891 non-null    bool   
 11  Southampton  891 non-null    bool   
 12  Cherbourg    891 non-null    bool   
dtypes: bool(3), float64(2), int64(4), object(4)
memory usage: 79.2+ KB


# Scale the numerical fields

In [561]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()

In [562]:
numerical_fields = ['Survived', 'Pclass', 'Age', 'SibSp', 'Parch', 'Fare']
scaler.fit(df[numerical_fields])

MinMaxScaler()

In [563]:
df[numerical_fields] = scaler.transform(df[numerical_fields])

In [564]:
df

Survived  Pclass       Age  SibSp     Parch      Fare  IsMrs   IsMr  \
0         0.0     1.0  0.271174  0.125  0.000000  0.014151  False   True   
1         0.0     1.0  0.434531  0.000  0.000000  0.015713  False   True   
2         0.0     1.0  0.401459  0.000  0.000000  0.016510  False   True   
3         0.0     0.0  0.673285  0.000  0.000000  0.101229  False   True   
4         0.0     1.0  0.246042  0.000  0.000000  0.015713  False   True   
..        ...     ...       ...    ...       ...       ...    ...    ...   
886       0.0     0.0  0.527326  0.000  0.000000  0.077294  False  False   
887       1.0     0.0  0.610455  0.000  0.000000  0.050610  False  False   
888       0.0     0.0  0.472229  0.000  0.000000  0.000000  False  False   
889       0.0     0.5  0.346569  0.000  0.166667  0.064412  False  False   
890       0.0     0.5  0.334004  0.000  0.000000  0.025374  False  False   

    IsMiss IsMaster  IsMale  Southampton  Cherbourg  
0    False    False   False         True      False  
1    False    False   False         True      False  
2    False    False   False        False      False  
3    False    False   False         True      False  
4    False    False   False         True      False  
..     ...      ...     ...          ...        ...  
886  False    False   False        False       True  
887  False    False   False         True      False  
888  False    False   False         True      False  
889  False    False   False         True      False  
890  False    False   False         True      False  

[891 rows x 13 columns]

# Data now ready for the model

In [565]:
from sklearn.model_selection import train_test_split

In [566]:
X_train, X_test, y_train, y_test = train_test_split(df.drop('Survived',axis=1), 
                                                    df['Survived'], test_size=0.30, 
                                                    random_state=101)

In [567]:
logmodel = LogisticRegression()
logmodel.fit(X_train,y_train)

LogisticRegression()

In [568]:
predictions = logmodel.predict(X_test)

In [569]:
from sklearn.metrics import classification_report

In [570]:
print (classification_report(y_test, predictions))

              precision    recall  f1-score   support

         0.0       0.82      0.85      0.84       171
         1.0       0.72      0.68      0.70        97

    accuracy                           0.79       268
   macro avg       0.77      0.76      0.77       268
weighted avg       0.79      0.79      0.79       268

